File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/Graph/allPathsSourceTarget_797.java
package algo.tzashinorpu.SpecialSubject.Graph;

import java.util.ArrayList;
import java.util.List;

public class allPathsSourceTarget_797 {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<Integer> path = new ArrayList<>();
        traverse(graph, 0, path);
        return res;
    }

    private void traverse(int[][] graph, int s, List<Integer> path) {
        path.add(s);
        int n = graph.length;
        if (s == n - 1) {
            res.add(new ArrayList<>(path));
        }
        for (int v : graph[s]) {
            traverse(graph, v, path);
        }
        path.remove(path.size() - 1);
    }
}

File: Clion/ZeroTrac/Graph/allPathsSourceTarget_797.cpp
#include "../header/g"

using namespace G;

vector<vector<int>> Solution::allPathsSourceTarget(vector<vector<int>> &graph)
{
    vector<vector<int>> res;
    int n = graph.size();
    vector<int> path;
    function<void(vector<int> &, int, vector<vector<int>> &)> dfs =
        [&](vector<int> &path, int v, vector<vector<int>> &g) -> void
    {
        path.push_back(v);
        if (v == n - 1)
            res.push_back(path);
        for (auto &s : graph[v])
            dfs(path, s, graph);
        path.pop_back();
    };
    dfs(path, 0, graph);
    return res;
}